### ISSUE TITLE
`torch.jit.trace` memory usage increase although forward is constant, and gets much slower than forward with model depth increase

### ISSUE THREAD

**Issue by fxmarty** [pytorch/pytorch#93943]
### 🐛 Describe the bug

Using `traced_model = torch.jit.trace(model, example_inputs)` is increasingly slower than `result = model(example_inputs)` over the model depth, and `torch.jit.trace` memory usage increases with model depth although `model.forward` is constant.

The model in question has a loop construct where a submodule is called iteratively.

Note that for both running `forward` and `torch.jit.trace`, the decorator `torch.inference_mode()` is used.

The results are:

<details>
  <summary>Slowdown of `torch.jit.trace`</summary>

![image](https://user-images.githubusercontent.com/9808326/216381415-13a92b88-c71c-4479-88bd-bc858a47414a.png)

![image](https://user-images.githubusercontent.com/9808326/216381500-3a4be27f-b264-4f62-8bb3-cf7102bb979f.png)

</details>

<details>
  <summary>Memory usage of `torch.jit.trace`</summary>

* n_loop = 30: max allocation `5270 MB`, `max_mem/n_loop = 175.66`
![image](https://user-images.githubusercontent.com/9808326/216382244-6b62e308-c97c-47f6-bd89-00960640a804.png)

* n_loop = 40: max allocation `6592 MB`, `max_mem/n_loop = 164.8`
![image](https://user-images.githubusercontent.com/9808326/216382316-dc57fe01-8c1c-4be4-a08a-fbdc76073053.png)

* n_loop = 50: max allocation `8084 MB`, `max_mem/n_loop = 161.68`
![image](https://user-images.githubusercontent.com/9808326/216382379-6d0b678d-9516-4a8a-9cfb-fe42965c094d.png)

* n_loop = 60: max allocation `9532 MB`, `max_mem/n_loop = 158.86`
![image](https://user-images.githubusercontent.com/9808326/216382453-09d3b977-b9f3-40d8-a200-2655920450aa.png)

* n_loop = 120: max allocation `18181 MB`, `max_mem/n_loop = 151.50`
![image](https://user-images.githubusercontent.com/9808326/216382523-178144a3-65e7-49ef-a3e9-cb5150ebd09f.png)

</details>

The loop looks like this: https://github.com/fxmarty/optimum/blob/9ef352b0f1775fe5e0048490722ea931aa6f4fd5/research/stable_diffusion_end_to_end_onnx/scriptable_pipeline_stable_diffusion.py#L298-L318

The script to run the forward is: https://github.com/fxmarty/optimum/blob/9ef352b0f1775fe5e0048490722ea931aa6f4fd5/research/stable_diffusion_end_to_end_onnx/run_custom_pytorch_pipeline.py

The script to run `torch.jit.trace` is: https://github.com/fxmarty/optimum/blob/9ef352b0f1775fe5e0048490722ea931aa6f4fd5/research/stable_diffusion_end_to_end_onnx/create_scriptmodule_by_tracing.py

### Versions

PyTorch version: 1.13.1+cu117
Is debug build: False
CUDA used to build PyTorch: 11.7
ROCM used to build PyTorch: N/A

OS: Ubuntu 22.04.1 LTS (x86_64)
GCC version: (Ubuntu 11.3.0-1ubuntu1~22.04) 11.3.0
Clang version: 14.0.0-1ubuntu1
CMake version: version 3.25.0
Libc version: glibc-2.35

Python version: 3.9.12 (main, Apr  5 2022, 06:56:58)  [GCC 7.5.0] (64-bit runtime)
Python platform: Linux-5.15.0-58-generic-x86_64-with-glibc2.35
Is CUDA available: True
CUDA runtime version: 11.7.99
CUDA_MODULE_LOADING set to: LAZY
GPU models and configuration: GPU 0: NVIDIA GeForce RTX 3060 Laptop GPU
Nvidia driver version: 515.86.01
cuDNN version: Probably one of the following:
/usr/local/cuda-11.7/targets/x86_64-linux/lib/libcudnn.so.8.7.0
/usr/local/cuda-11.7/targets/x86_64-linux/lib/libcudnn_adv_infer.so.8.7.0
/usr/local/cuda-11.7/targets/x86_64-linux/lib/libcudnn_adv_train.so.8.7.0
/usr/local/cuda-11.7/targets/x86_64-linux/lib/libcudnn_cnn_infer.so.8.7.0
/usr/local/cuda-11.7/targets/x86_64-linux/lib/libcudnn_cnn_train.so.8.7.0
/usr/local/cuda-11.7/targets/x86_64-linux/lib/libcudnn_ops_infer.so.8.7.0
/usr/local/cuda-11.7/targets/x86_64-linux/lib/libcudnn_ops_train.so.8.7.0
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

Versions of relevant libraries:
[pip3] mypy-extensions==0.4.3
[pip3] numpy==1.22.4
[pip3] torch==1.13.1
[pip3] torch-model-archiver==0.6.1
[pip3] torch-workflow-archiver==0.2.5
[pip3] torchinfo==1.7.0
[pip3] torchserve==0.6.1
[pip3] torchtriton==2.0.0+0d7e753227
[pip3] torchvision==0.14.1
[conda] cudatoolkit               11.3.1               h2bc3f7f_2    anaconda
[conda] numpy                     1.22.4                   pypi_0    pypi
[conda] torch                     1.13.1                   pypi_0    pypi
[conda] torch-model-archiver      0.6.1                    pypi_0    pypi
[conda] torch-workflow-archiver   0.2.5                    pypi_0    pypi
[conda] torchinfo                 1.7.0                    pypi_0    pypi
[conda] torchserve                0.6.1                    pypi_0    pypi
[conda] torchtriton               2.0.0+0d7e753227          pypi_0    pypi
[conda] torchvision               0.14.1                   pypi_0    pypi

**Comment by fxmarty** [comment]
During tracing, the RAM and CU usage is as follow:

1. Maximum CPU usage, no large memory allocation
2. Low CPU usage (single core), no large memory allocation 
3. Maximum CPU usage, first allocation cliff (larger when number of loops higher)
4. Maximum CPU usage, no large memory allocation 
5. Low CPU usage (single core), no large memory allocation
6. Maximum CPU usage, second allocation cliff (larger when number of loops higher)

![image](https://user-images.githubusercontent.com/9808326/216393820-410fe8df-644e-47af-87c3-df758f324b58.png)

**Comment by fxmarty** [comment]
It appears `torch.jit.trace` does not reuse variables in the loop:

```
  _72 = (unet).forward(torch.cat([latents, latents]), _11, _9, )
  noise_pred0, = _72
  latents0 = torch.slice(noise_pred0, 0, 0, 1)
  _73 = (unet).forward(torch.cat([latents0, latents0]), _12, _9, )
  noise_pred1, = _73
  latents1 = torch.slice(noise_pred1, 0, 0, 1)
  _74 = (unet).forward(torch.cat([latents1, latents1]), _13, _9, )
  noise_pred2, = _74
  latents2 = torch.slice(noise_pred2, 0, 0, 1)
  _75 = (unet).forward(torch.cat([latents2, latents2]), _14, _9, )
  noise_pred3, = _75
  latents3 = torch.slice(noise_pred3, 0, 0, 1)
  ...
```

So I suspect it is all those intermediate dead variables that accumulate memory?

**Comment by sanchitintel** [comment]
Duplicate of #35600

**Comment by fxmarty** [comment]
@sanchitintel Is it really? In #35600 the jit.trace is call in the loop, although in my case jit.trace is called only once and there is a loop in the model.

**Comment by un-certainty** [comment]
> @sanchitintel Is it really? In #35600 the jit.trace is call in the loop, although in my case jit.trace is called only once and there is a loop in the model.

@sanchitintel @fxmarty 

This issue is also related: https://github.com/onnx/onnx/issues/4319

**Comment by sanchitintel** [comment]
Hello @fxmarty & @un-certainty, I'm sorry for having missed your replies!
Yes, this issue is a duplicate of #35600 because the memory-leak issue is unrelated to using `torch.jit.trace` in a loop - it's simply the case that the bug was observed in the original issue when memory usage became too high due to multiple `torch.jit.trace` calls.

#96726 has added some more debug info.

**Comment by chengzeyi** [comment]
> Hello @fxmarty & @un-certainty, I'm sorry for having missed your replies! Yes, this issue is a duplicate of #35600 because the memory-leak issue is unrelated to using `torch.jit.trace` in a loop - it's simply the case that the bug was observed in the original issue when memory usage became too high due to multiple `torch.jit.trace` calls.
> 
> #96726 has added some more debug info.

Hi, please see my comment in this issue:
[https://github.com/pytorch/pytorch/issues/96726#issuecomment-1709868700](url)

**Comment by janeyx99** [comment]
TorchScript has been deprecated in favor of torch.export as of 2.11. https://pytorch.org/blog/pytorch-2-11-release-blog/ Closing this issue as we are not further developing on it--please ping if you believe this issue was erroneously closed.
